### PR TITLE
feat: multiple popups in a hierarchy-like system

### PR DESCRIPTION
**Pull request by Rukario**
Large popup like inspector page can stick while a new popup lesser important to it can open to co-exist and dismiss even lesser popups if there's any, the system works like hierarchy, lesser important ones go away before others.

Transmission popup hierarchy

Notes: Implemented a new popup management system for web client to support multiple popups in a hierarchy-like system.

### CODE REVIEW

**Review by ckerr — APPROVED** [review]
UI and code LGTM.

I'm a little concerned about the added popup complexity, but I don't think it really hurts anything with our current popups :man_shrugging:

**Comment by Rukario** [comment]
That's the nature of adding a feature that enable things to do multitasking.

File: web/src/transmission.js [hunk 1]
```diff
@@ -31,8 +31,10 @@ export class Transmission extends EventTarget {
 
     // Initialize the helper classes
     this.action_manager = action_manager;
+    this.max_popups = 2;
```
Comment (tearfur) [inline web/src/transmission.js:34]: To ensure this value is not accidentally modified.

```js
static get max_popups() {
  return 2;
}
```

You can optionally define another property for convenience:

```js
static get default_popup_level() {
  return Transmission.max_popups - 1;
}
```

File: web/src/transmission.js [hunk 2]
```diff
@@ -125,18 +125,18 @@ export class Transmission extends EventTarget {
...
           }
           break;
         case 'show-move-dialog':
           this.setCurrentPopup(new MoveDialog(this, this.remote));
           break;
         case 'show-overflow-menu':
-          if (this.popup instanceof OverflowMenu) {
-            this.setCurrentPopup(null);
+          if (this.popup[1] instanceof OverflowMenu) {
+            this.popup[1].close();
```
Comment (tearfur) [inline web/src/transmission.js:139]:
```suggestion
          if (this.popup[Transmission.default_popup_level] instanceof OverflowMenu) {
            this.popup[Transmission.default_popup_level].close();
```

File: web/src/transmission.js [hunk 3]
```diff
@@ -1157,23 +1157,25 @@ TODO: fix this when notifications get fixed
...
-      throw new Error(event_);
-    }
-    this.popup.removeEventListener('close', this.boundPopupCloseListener);
-    delete this.popup;
-  }
-
-  setCurrentPopup(popup) {
-    if (this.popup) {
-      this.popup.close();
+  setCurrentPopup(popup, level = this.max_popups - 1) {
```
Comment (tearfur) [inline web/src/transmission.js:1160]:
```suggestion
  setCurrentPopup(popup, level = Transmission.default_popup_level) {
```

File: web/src/transmission.js [hunk 4]
```diff
@@ -777,7 +777,7 @@ TODO: fix this when notifications get fixed
     const meta_key = event_.metaKey || event_.ctrlKey,
       { row } = event_.currentTarget;
 
-    if (this.popup && this.popup.name !== 'inspector') {
+    if (this.popup[this.max_popups - 1]) {
```
Comment (tearfur) [inline web/src/transmission.js:780]:
```suggestion
    if (this.popup[Transmission.default_popup_level]) {
```

File: web/src/transmission.js [hunk 5]
```diff
@@ -278,8 +278,8 @@ export class Transmission extends EventTarget {
     } else {
       this.elements.torrent_list.addEventListener('contextmenu', (event_) => {
         rightc(event_);
-        if (this.popup) {
-          this.popup.root.style.pointerEvents = 'auto';
+        if (this.popup[1]) {
+          this.popup[1].root.style.pointerEvents = 'auto';
```
Comment (tearfur) [inline web/src/transmission.js:282]: ditto

File: web/src/transmission.js [hunk 6]
```diff
@@ -263,8 +263,8 @@ export class Transmission extends EventTarget {
         clearTimeout(this.busyclick);
         this.busyclick = false;
         setTimeout(() => {
-          if (this.popup) {
-            this.popup.root.style.pointerEvents = 'auto';
+          if (this.popup[1]) {
+            this.popup[1].root.style.pointerEvents = 'auto';
```
Comment (tearfur) [inline web/src/transmission.js:267]: ditto

File: web/src/transmission.js [hunk 7]
```diff
@@ -195,15 +195,15 @@ export class Transmission extends EventTarget {
     document.addEventListener('keyup', this._keyUp.bind(this));
     e = document.querySelector('#torrent-container');
     e.addEventListener('click', (e_) => {
-      if (this.popup && this.popup.name !== 'inspector') {
+      if (this.popup[this.max_popups - 1]) {
```
Comment (tearfur) [inline web/src/transmission.js:198]: ditto

**Review by tearfur — COMMENTED** [review]
I'm late to the party, but still I have some points I want to raise.

File: web/src/transmission.js [hunk 8]
```diff
@@ -1157,23 +1157,25 @@ TODO: fix this when notifications get fixed
...
-    this.popup.removeEventListener('close', this.boundPopupCloseListener);
-    delete this.popup;
-  }
-
-  setCurrentPopup(popup) {
-    if (this.popup) {
-      this.popup.close();
+  setCurrentPopup(popup, level = this.max_popups - 1) {
+    let index = level;
+    while (index < this.max_popups) {
```
Comment (tearfur) [inline web/src/transmission.js:1162]: This should be a for loop instead.

**Comment by Rukario** [comment]
@tearfur Your suggestions are pretty good, looks like a call for a refactor PR.

**Pull request merged**